### PR TITLE
feat(ast)!: add `IdentifierReference` to `ExportSpecifier`

### LINKED ISSUE #3795: Didn't catch the reference in `ExportNamedDeclaration`
The references of `A` is empty for:

```typescript
import { A } from 'foo'

export  { A }
```
CleanShot 2024-06-20 at 19 39 38@2x

[playground](https://oxc-project.github.io/oxc/playground/?code=3YCAAIClgICAgICAgIC0G8rnONK89Iyo%2FIRvUICdKXoMpBKY%2Bxn64LMYJ7DwPsN%2Ff9OYgIA%3D)

### LINKED ISSUE #3796: semantic: should not depend on module record
babel exports/imported-types/input.ts

```javascript
import type A from "A";
import type { B } from "B";
import C from "C";

export { A, B, C } // <-- A and B will be removed
```

is transformed to

```
export { C };
```

after `ModuleRecord` is removed, `add_export_flag_for_export_identifier` should not depend on `ModulreRecord`.

### PR DESCRIPTION
**Pull request by Boshen**
closes #3795
closes #3796

```
import { A } from 'foo'

export  { A }
          ^ This is now an `IdentifierReference` instead of `IdentifierName` per spec.
```

### CODE REVIEW

**Comment by Boshen** [comment]
* **#3820** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3820?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Boshen and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_semantic/src/builder.rs [hunk 1]
```diff
@@ -388,37 +384,6 @@ impl<'a> SemanticBuilder<'a> {
     pub fn add_redeclare_variable(&mut self, symbol_id: SymbolId, span: Span) {
         self.symbols.add_redeclare_variable(symbol_id, span);
     }
-
-    fn add_export_flag_for_export_identifier(&mut self) {
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:392]: It seems like we are lacking tests for adding export flags.

File: crates/oxc_ast/src/ast/js.rs [hunk 2]
```diff
@@ -1955,6 +1955,8 @@ pub enum ExportDefaultDeclarationKind<'a> {
 #[cfg_attr(feature = "serialize", derive(Serialize, Tsify))]
 #[cfg_attr(feature = "serialize", serde(untagged))]
 pub enum ModuleExportName<'a> {
-    Identifier(IdentifierName<'a>),
+    IdentifierName(IdentifierName<'a>),
```
Comment (Dunqing) [inline crates/oxc_ast/src/ast/js.rs:1958]: Where's the need for this?

Comment (Boshen) [reply crates/oxc_ast/src/ast/js.rs:1958]: All other usages are still `IdentifierName`.

Comment (Dunqing) [reply crates/oxc_ast/src/ast/js.rs:1958]: I think we should just need `IdentifierReference`. But it may require a lot of changes.

**Pull request merged**